// ngx-scrollspy/dist/core/element.directive.js
import { Directive, Injectable, Input } from '@angular/core';
import { ReplaySubject } from 'rxjs/ReplaySubject';
import { ScrollSpyService } from './service';
var ScrollSpyElementDirective = (function () {
    function ScrollSpyElementDirective(scrollSpy) {
        this.scrollSpy = scrollSpy;
        this.scrollStream$ = new ReplaySubject(1);
    }
    ScrollSpyElementDirective.prototype.ngOnInit = function () {
        if (!this.scrollSpyId) {
            return console.warn('ScrollSpy: Missing id.');
        }
        if (!!this.scrollSpy.getObservable(this.scrollSpyId)) {
            console.warn('ScrollSpy: duplicate id "' + this.scrollSpyId + '". Instance will be skipped!');
        }
        else {
            this.scrollSpy.setObservable(this.scrollSpyId, this.scrollStream$);
        }
    };
    ScrollSpyElementDirective.prototype.ngOnDestroy = function () {
        this.scrollSpy.deleteObservable(this.scrollSpyId);
    };
    ScrollSpyElementDirective.prototype.onScroll = function ($event) {
        this.scrollStream$.next($event);
    };
    return ScrollSpyElementDirective;
}());
export { ScrollSpyElementDirective };
ScrollSpyElementDirective.decorators = [
    { type: Injectable },
    { type: Directive, args: [{
                selector: '[scrollSpyElement]',
                host: {
                    '(scroll)': 'onScroll($event)'
                }
            },] },
];
/** @nocollapse */
ScrollSpyElementDirective.ctorParameters = function () { return [
    { type: ScrollSpyService, },
]; };
ScrollSpyElementDirective.propDecorators = {
    'scrollSpyId': [{ type: Input, args: ['scrollSpyElement',] },],
};
