// ngx-scrollspy/dist/core/service.js
import { Injectable, EventEmitter } from '@angular/core';
var ScrollSpyService = (function () {
    function ScrollSpyService() {
        this.changes$ = new EventEmitter();
        this.observables = {};
    }
    ScrollSpyService.prototype.getObservable = function (key) {
        return this.observables[key];
    };
    ScrollSpyService.prototype.setObservable = function (key, observable) {
        this.observables[key] = observable;
        this.observables = this.observables = Object.assign({}, this.observables);
        this.changes$.emit({ index: key, change: 'set' });
    };
    ScrollSpyService.prototype.deleteObservable = function (key) {
        delete this.observables[key];
        this.observables = Object.assign({}, this.observables);
        this.changes$.emit({ index: key, change: 'delete' });
    };
    return ScrollSpyService;
}());
export { ScrollSpyService };
ScrollSpyService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
ScrollSpyService.ctorParameters = function () { return []; };
