// ngx-scrollspy/dist/core/window.directive.js
import { Directive, Injectable } from '@angular/core';
import { ReplaySubject } from 'rxjs/ReplaySubject';
import { ScrollSpyService } from './service';
var ScrollSpyDirective = (function () {
    function ScrollSpyDirective(scrollSpy) {
        this.scrollSpy = scrollSpy;
        this.scrollStream$ = new ReplaySubject(1);
    }
    ScrollSpyDirective.prototype.ngOnInit = function () {
        if (!!this.scrollSpy.getObservable('window')) {
            console.warn('ScrollSpy: duplicate id "window". Instance will be skipped!');
        }
        else {
            this.scrollSpy.setObservable('window', this.scrollStream$);
        }
    };
    ScrollSpyDirective.prototype.onScroll = function ($event) {
        this.scrollStream$.next($event);
    };
    return ScrollSpyDirective;
}());
export { ScrollSpyDirective };
ScrollSpyDirective.decorators = [
    { type: Injectable },
    { type: Directive, args: [{
                selector: '[scrollSpy]',
                host: {
                    '(window:scroll)': 'onScroll($event)'
                }
            },] },
];
/** @nocollapse */
ScrollSpyDirective.ctorParameters = function () { return [
    { type: ScrollSpyService, },
]; };
