// ngx-scrollspy/dist/plugin/index.directive.js
import { Directive, Injectable, ElementRef, Input } from '@angular/core';
import { ScrollSpyIndexService } from './index.service';
var ScrollSpyIndexDirective = (function () {
    function ScrollSpyIndexDirective(elRef, scrollSpyIndex) {
        this.elRef = elRef;
        this.scrollSpyIndex = scrollSpyIndex;
        this.defaultOptions = {
            selector: 'anchor'
        };
        this.el = elRef.nativeElement;
    }
    ScrollSpyIndexDirective.prototype.ngOnInit = function () {
        if (!this.options) {
            this.options = {};
        }
        if (!this.options.id) {
            return console.warn('ScrollSpyIndex: Missing id.');
        }
        this.options = Object.assign(this.defaultOptions, this.options);
    };
    ScrollSpyIndexDirective.prototype.ngAfterViewInit = function () {
        this.scrollSpyIndex.setIndex(this.options.id, this.el.getElementsByClassName(this.options.selector));
    };
    ScrollSpyIndexDirective.prototype.ngOnDestroy = function () {
        this.scrollSpyIndex.deleteIndex(this.options.id);
    };
    return ScrollSpyIndexDirective;
}());
export { ScrollSpyIndexDirective };
ScrollSpyIndexDirective.decorators = [
    { type: Injectable },
    { type: Directive, args: [{
                selector: '[scrollSpyIndex]'
            },] },
];
/** @nocollapse */
ScrollSpyIndexDirective.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: ScrollSpyIndexService, },
]; };
ScrollSpyIndexDirective.propDecorators = {
    'options': [{ type: Input, args: ['scrollSpyIndex',] },],
};
