// ngx-scrollspy/dist/plugin/index.service.js
import { Injectable, EventEmitter } from '@angular/core';
var ScrollSpyIndexService = (function () {
    function ScrollSpyIndexService() {
        this.changes$ = new EventEmitter();
        this.indexes = {};
    }
    ScrollSpyIndexService.prototype.getIndex = function (key) {
        return this.indexes[key];
    };
    ScrollSpyIndexService.prototype.setIndex = function (key, index) {
        this.indexes[key] = index;
        this.indexes = this.indexes = Object.assign({}, this.indexes);
        this.changes$.emit({ index: key, change: 'set' });
    };
    ScrollSpyIndexService.prototype.deleteIndex = function (key) {
        delete this.indexes[key];
        this.indexes = Object.assign({}, this.indexes);
        this.changes$.emit({ index: key, change: 'delete' });
    };
    return ScrollSpyIndexService;
}());
export { ScrollSpyIndexService };
ScrollSpyIndexService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
ScrollSpyIndexService.ctorParameters = function () { return []; };
